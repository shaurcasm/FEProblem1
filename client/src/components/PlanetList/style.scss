@import '../../app/colors.scss';

.sr-only {
    position: absolute;
    left: -1000px;
    top: auto;
    width: 1px;
    height: 1px;
    overflow: hidden;
}

.planet-list {
    margin: 1.5vh 0;

    input {
        background-color: $cultured;
        border-color: $space-cadet;
        width: 150px;
        height: 30px;
        font-size: 1.2rem;
        color: $space-cadet;
        text-indent: 1vw;
    }
}