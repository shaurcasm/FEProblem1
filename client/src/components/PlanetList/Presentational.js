import React, { useEffect, useState } from 'react';
import './style.scss';

// Any datalist styling; 
const Presentational = ({ planets, select, changeImage }) => {
    const [selectedPlanet, setSelection] = useState('');
    const [previousSelection, setPrevious] = useState('');

    // Valuable code. Validating Datalist input.
    // When the value of the input changes...
    const inputValidation = event => {
        var optionFound = false,
            input = event.target,
            datalist = input.list;
        
        // Determine whether an option exists with the current value of the input.
        for (let j = 0; j < datalist.options.length; j++) {
            let optionValue = datalist.options[j].value;
            if(input.value.toUpperCase() === optionValue.toUpperCase()) {
                optionFound = true;
                setSelection(optionValue);
                changeImage(optionValue);
                break;
            }
        }
        // Use the setCustomValidity function of the validation API
        // to provide a user feedback if the value does not exist in the datalist.
        if(optionFound) {
            input.setCustomValidity('');
        } else {
            input.setCustomValidity('Please select a valid Planet');
        }
        input.reportValidity();
    }

    // Apply if local state selected planet is changed(second argument)
    useEffect(() => {
        //console.log("Current Planet: ", selectedPlanet);
        //console.log("Previous Planet: ", previousSelection);

        if(selectedPlanet !== previousSelection) {
            // If there was a previous selection, replace it with new one use redux action
            if(previousSelection) {
                //console.log('Replace action!');
                select.replacePlanet(selectedPlanet, previousSelection);
                setPrevious(selectedPlanet);
            }
            // Else, no previous selection. Merely use add action.
            else {
                //console.log('Add action!')
                select.addPlanet(selectedPlanet);
                setPrevious(selectedPlanet);
            }
        }
    }, [previousSelection, select, selectedPlanet]);

    // Gets updated everytime planets state is changed.
    const planetOptions = planets.filter(planet => planet.selected === false).map((planet, index) => {
        return (
            <option key={index} value={planet.name}>{planet.name}</option>
        )        
    });
    
    // Comment: PlanetList component's datalist didn't need dynamic ID to function properly, for currently unknown reason.
    // Unlike VehicleList, yet they both were applied once by multiple sectional components.
    return (
        <div className='planet-list'>
            <label htmlFor='planet-selector' className='sr-only'>Choose a Planet from this list:</label>
            <input type='text' list='available-planets' name='planet-selector' onChange={inputValidation} placeholder="Pick a Planet." autoCorrect='off' spellCheck='false' autoFocus />
            <datalist id='available-planets'>
                {planetOptions}
            </datalist>
        </div>
    );
}

export default Presentational;