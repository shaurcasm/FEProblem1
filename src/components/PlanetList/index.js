import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import inputValidation from '../../utilities/inputValidation.js';
import './style.scss';

// Any datalist styling; 
const PlanetList = ({ planets, select, changeImage }) => {
    const [selectedPlanet, setSelection] = useState('');
    const [previousSelection, setPrevious] = useState('');

    // Apply if local state selected planet is changed(second argument)
    useEffect(() => {
        if (selectedPlanet === previousSelection)
            return;

        // If there was a previous selection, replace it with new one use redux action
        if (previousSelection) {
            select.replacePlanet(selectedPlanet, previousSelection);
            setPrevious(selectedPlanet);
        }
        // Else, no previous selection. Merely use add action.
        else {
            select.addPlanet(selectedPlanet);
            setPrevious(selectedPlanet);
        }
    }, [previousSelection, select, selectedPlanet]);

    const validate = event => inputValidation(event, planets, setSelection, changeImage, null);

    // Gets updated everytime planets state is changed.
    const planetOptions = planets.filter(planet => planet.selected === false).map((planet, index) => {
        return (
            <option key={index} value={planet.name}>
                {planet.name}
            </option>
        )
    });

    return (
        <div className='planet-list'>
            <label htmlFor='planet-selector' className='sr-only'>
                Choose a Planet from this list:
            </label>
            <input 
                type='text'
                list='available-planets'
                name='planet-selector'
                onChange={validate}
                placeholder="Pick a Planet"
                autoCorrect='off'
                spellCheck='false'
                autoFocus
            />
            <datalist id='available-planets'>
                {planetOptions}
            </datalist>
        </div>
    );
}

PlanetList.propTypes = {
    planets: PropTypes.array,
    select: PropTypes.objectOf(PropTypes.func),
    changeImage: PropTypes.func
}

export default PlanetList;