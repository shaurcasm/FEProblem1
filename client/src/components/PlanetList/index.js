import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import * as selectPlanets from '../../actions/selectPlanets.js'
import Presentational from './Presentational.js'

const mapStateToProps = state => ({
    planets: state.planetList.planetArray
});

const mapDispatchToProps = dispatch => ({
    select: bindActionCreators(selectPlanets, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(Presentational);