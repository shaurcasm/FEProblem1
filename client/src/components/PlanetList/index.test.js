import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

// Component to be tested
import PlanetList from './index.js';

const testSelect = {
    addPlanet: (planet) => console.log(planet),
    replacePlanet: (planetOne, planetTwo) => console.log('New Planet = ' + planetOne, 'Previous Planet = ' + planetTwo),
    resetPlanet: () => console.log('Reset')
};

const testChangeImage = (name) => console.log(name)

describe('<PlanetList />', () => {

    test('Renders the component', () => {
        
        const testPlanets = [
            { name: 'Mercury', selected: false },
            { name: 'Venus', selected: false },
            { name: 'Earth', selected: false },
            { name: 'Mars', selected: false }
        ];
        const component = shallow(<PlanetList planets={testPlanets} select={testSelect} changeImage={testChangeImage} />);
        expect(toJson(component)).toMatchSnapshot();
    });

    describe('Datalist', () => {
        
        test('Validate Datalist with 3 Planets', () => {

            const testPlanets = [
                { name: 'Mercury', selected: false },
                { name: 'Venus', selected: false },
                { name: 'Earth', selected: false }
            ];
            const component = shallow(<PlanetList planets={testPlanets} select={testSelect} changeImage={testChangeImage} />);

            const datalist = component.find('#available-planets');
            expect(datalist.children().length).toEqual(testPlanets.length);
        });

        
        test('Validate Datalist with 4 Planets', () => {

            const testPlanets = [
                { name: 'Mercury', selected: false },
                { name: 'Venus', selected: false },
                { name: 'Earth', selected: false },
                { name: 'Mars', selected: false }
            ];
            const component = shallow(<PlanetList planets={testPlanets} select={testSelect} changeImage={testChangeImage} />);

            const datalist = component.find('#available-planets');
            expect(datalist.children().length).toEqual(testPlanets.length);
        });
    });
});