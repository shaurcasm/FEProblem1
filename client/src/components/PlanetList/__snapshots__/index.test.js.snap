// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlanetList /> Renders the component 1`] = `
<div
  className="planet-list"
>
  <label
    className="sr-only"
    htmlFor="planet-selector"
  >
    Choose a Planet from this list:
  </label>
  <input
    autoCorrect="off"
    autoFocus={true}
    list="available-planets"
    name="planet-selector"
    onChange={[Function]}
    placeholder="Pick a Planet."
    spellCheck="false"
    type="text"
  />
  <datalist
    id="available-planets"
  >
    <option
      key="0"
      value="Mercury"
    >
      Mercury
    </option>
    <option
      key="1"
      value="Venus"
    >
      Venus
    </option>
    <option
      key="2"
      value="Earth"
    >
      Earth
    </option>
    <option
      key="3"
      value="Mars"
    >
      Mars
    </option>
  </datalist>
</div>
`;
